File: src/cc/linkedlist/LinkedListCycle.cc
#include <cstdlib>
#include <unordered_set>

using namespace std;

/**
 * Definition for singly-linked list.
 */
class ListNode {
public:
    int val;
    ListNode *next;
    ListNode(): val(0), next(NULL) {
    }
    ListNode(int x): val(x), next(NULL) {
    }
    ListNode(int x, ListNode *next): val(x), next(next) {
    }
};

/**
 * LeetCode 141. Linked List Cycle
 * https://leetcode.com/problems/linked-list-cycle/
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == NULL) {
            return false;
        }
        // return hasCycleHashing(head);
        return hasCycleTwoPointer(head);
    }
private:
    bool hasCycleHashing(ListNode *head) {
        unordered_set<ListNode*> aSet;
        ListNode *currentNode = head;
        while (currentNode != NULL) {
            if (aSet.find(currentNode) != aSet.end()) {
                return true;
            }
            aSet.insert(currentNode);
            currentNode = currentNode -> next;
        }
        return false;
    }
    bool hasCycleTwoPointer(ListNode *head) {
        ListNode *pFast = head;
        ListNode *pSlow = head;
        while (pFast != NULL && pFast -> next != NULL) {
            pFast = pFast -> next -> next;
            pSlow = pSlow -> next;
            if (pFast == pSlow) {
                return true;
            }
        }
        return false;
    }
};

int main(int argc, char const *argv[]) {
    // ...
    return 0;
}
/* EOF */

File: src/io/ziheng/list/leetcode/LinkedListCycle.java
package io.ziheng.list.leetcode;

import java.util.HashSet;
import java.util.Set;

/**
 * LeetCode 141. Linked List Cycle
 * https://leetcode.com/problems/linked-list-cycle/
 */
public class LinkedListCycle {
    public boolean hasCycle(ListNode head) {
        if (head == null) {
            return false;
        }
        // return hasCycleHashing(head);
        return hasCycleTwoPointer(head);
    }
    /**
     * Linked List has Cycle -> Two Pointer
     *
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     *
     * @param head
     * @return boolean
     */
    private boolean hasCycleTwoPointer(ListNode head) {
        ListNode pFast = head;
        ListNode pSlow = head;
        while (pFast != null && pFast.next != null) {
            pFast = pFast.next.next;
            pSlow = pSlow.next;
            if (pFast == pSlow) {
                return true;
            }
        }
        return false;
    }
    /**
     * Linked List has Cycle -> Hashing
     *
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     *
     * @param head
     * @return boolean
     */
    private boolean hasCycleHashing(ListNode head) {
        Set<ListNode> aSet = new HashSet<>();
        ListNode currentNode = head;
        while (currentNode != null) {
            if (aSet.contains(currentNode)) {
                return true;
            }
            aSet.add(currentNode);
            currentNode = currentNode.next;
        }
        return false;
    }
}
/* EOF */